crypto/src/builder.rs

```diff
@@ -107,13 +118,40 @@ impl TransactionBuilder {
         self
     }
 
-    pub fn finalize<R: CryptoRng + RngCore>(mut self, rng: &mut R) -> Transaction {
-        // Randomize outputs to minimize info leakage.
-        self.outputs.shuffle(rng);
-        self.spends.shuffle(rng);
+    /// Set the expiry height.
+    pub fn set_expiry_height(mut self, expiry_height: u32) -> Self {
+        self.expiry_height = Some(expiry_height);
+        self
+    }
+
+    /// Set the chain ID.
+    pub fn set_chain_id(mut self, chain_id: String) -> Self {
+        self.chain_id = Some(chain_id);
+        self
+    }
+
+    pub fn finalize<R: CryptoRng + RngCore>(mut self, rng: &mut R) -> Result<Transaction, Error> {
+        // Randomize all actions (including outputs) to minimize info leakage.
+        self.actions.shuffle(rng);
+
+        if self.chain_id.is_none() {
+            return Err(Error::NoChainID);
+        }
+
+        if self.expiry_height.is_none() {
+            return Err(Error::ExpiryHeightNotSet);
+        }
+
+        if self.fee.is_none() {
+            return Err(Error::FeeNotSet);
+        }
```

Comment:
We should do this later, but I think it would probably be good to assert that the transaction balances as constructed, rather than waiting for its verification to fail.

We could do this by keeping a vec of `Value`s in the builder, then using a `BTreeMap` or something to sum up per-id amounts and checking balances.